### PR TITLE
Add Minor Features

### PR DESCRIPTION
**Pull request by dan-tang-ssd**
I added some minor features as below:

My Role page
 - This is available for institutional admin / assessor / member only. Site admin and site manager cannot access it because they do not belong to any institution.
 - It shows organisation name, user name, role name.
 - It shows a link to "Request to leave an institution"
 - It shows a link to "Request to remove everything for institution"

---

Request to leave an institution
  - Optional to remove personal data (just to overwrite users.name, users.email with dummy value)
  - Delete the corresponding organisation_members record
     - user still can login but unable to enter main menu because user does not belong to any institution now

---

Request to remove everything for institution
  - Accessible for institutional admin only
  - Once submit, a new removal_requests record will be created with below items:
     - Organisation id, name, requester id, name, email
     - These data values will be kept for future enquiry or reference after everything is removed for an institution
     - Nothing change after submitting this request, user can login and use data platform as usual
   
  - Note: Email template will be developed once confirmed email will be sent for which actions

P.S. Text content, UI look and feel to be revised

---

Removal Request CRUD panel
 - create / update / delete are disabled
 - preview is enabled
 - added custom buttons for Cancel, Remind, Confirm, Remove Everything
 - status:
    - REQUESTED: data removal request has been submitted by institutional admin
    - CANCELLED: site admin / site manager manually click Cancel button to cancel removal request
    - REMINDER_SENT: site admin / site manager manually click Remind button to send reminder email to requester
    - FINAL_CONFIRMED: site admin / site manager manually click Confirm button to final confirm data removal (Suppose requester should reply reminder email to site admin / site manager to final confirm data removal request)
    - EVERYTHING_REMOVED: site admin / site manager manually click Remove Everything button, all related records for an institution removed in database

Currently all operations are performed manually by site admin / site manager.
As the precise procedure is subject for discussion (this may take a long time). I have developed the core workflow and data removal programs only.

### CODE REVIEW

**Comment by dan-tang-ssd** [comment]
Here are some screen shots FYI:

My Role page
image

Request to leave an institution
image

Request to remove everything for institution
image

Removal Requests CRUD
image

**Comment by dan-tang-ssd** [comment]
Added email alert for data removal workflow:
 - Requested
 - Reminder (Both manual trigger and daily schedule job trigger)
 - Cancelled
 - Final Confirmed
 - Everything Removed

---

Below please find screen shots FYI:

Requested
image

Reminder
image

Cancelled
image

Final Confirmed
image

Everything Removed
image

**Review by dave-mills — CHANGES_REQUESTED** [review]
Looks good - I like the carefully considered set of options and functions surrounding the removal of either yourself or the entire institution from the platform. The set of options work well and seem pretty clear to me. 

>  Note: Email template will be developed once confirmed email will be sent for which actions
> P.S. Text content, UI look and feel to be revised

Yep, makes sense. Completely agree to make sure the functionality works and then review the email text, template + UI separately. 

> Currently all operations are performed manually by site admin / site manager.  
As the precise procedure is subject for discussion (this may take a long time). I have developed the core workflow and data removal programs only.

This is excellent. I think it's likely that they'll approve our recommended approach, but with this setup here, we can quickly adjust the final workflow based on any feedback they give. Good work!

Just one small request - I think we should add relationships to link the RemovalRequest model with organistaion and user, and then use that to show the org + requester details in the CRUD panel. Apart from that, this looks good to go.

File: .env.example [hunk 1]
```diff
@@ -33,3 +33,7 @@ MAILGUN_DOMAIN=
 MAILGUN_SECRET=
 MAIL_FROM_ADDRESS=no-reply@stats4sd.org
 MAIL_FROM_NAME="${APP_NAME}"
+
+# Define data removal email alert recipient in CSV format
+# It is recommended to define a email distribution list, so user can have full control without updating this configuration file
+DATA_REMOVAL_ALERT_RECIPIENTS=admin@example.com
```
[the remark this answers is not in the document]
Comment (dan-tang-ssd) [reply .env.example:39]: Yes, very likely.

File: resources/views/vendor/backpack/base/inc/menu.blade.php [hunk 2]
```diff
@@ -28,6 +28,13 @@
         <!-- Topbar. Contains the right part -->
         @include(backpack_view('inc.topbar_right_content'))
 
+        <!-- TODO: show My Role link if user is institutional members -->
+        @if ( auth()->user()?->hasRole('Institutional Admin') || auth()->user()?->hasRole('Institutional Assessor') || auth()->user()?->hasRole('Institutional Member') )
```
[the remark this answers is not in the document]
Comment (dan-tang-ssd) [reply resources/views/vendor/backpack/base/inc/menu.blade.php]: Thank you Dave, I have revised as suggested.

File: resources/views/vendor/backpack/base/inc/sidebar_content.blade.php [hunk 3]
```diff
@@ -140,3 +141,14 @@
 @if(Auth::user()->can('view project-level dashboard'))
     <li class="nav-item"><a class="nav-link" href="{{ backpack_url('generic-dashboard?level=initiative') }}"><i class="nav-icon la la-tachometer-alt"></i> Initiative Dashboard</a></li>
 @endif
+
+@if(Auth::user()->can('view institution-level dashboard') || Auth::user()->can('view portfolio-level dashboard') || Auth::user()->can('view project-level dashboard'))
```
[the remark this answers is not in the document]
Comment (dan-tang-ssd) [reply resources/views/vendor/backpack/base/inc/sidebar_content.blade.php]: Thank you Dave, I have revised as suggested.

File: database/migrations/2023_05_11_080031_create_removal_requests_table.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+     */
+    public function up(): void
+    {
+        Schema::create('removal_requests', function (Blueprint $table) {
+            $table->id();
+            $table->unsignedBigInteger('organisation_id');
+            $table->string('organisation_name');
+            $table->unsignedBigInteger('requester_id');
+            $table->string('requester_name');
+            $table->string('requester_email');
```
[the remark this answers is not in the document]
Comment (dan-tang-ssd) [reply database/migrations/2023_05_11_080031_create_removal_requests_table.php:20]: Yes, as the corresonding organisations and users records will be deleted. We will have no model to refer to after "Everything Removed". So I have to copy organisation name, requester name and email when data removal request is submitted. These values will be kept after "Everything Removed" for possible enquiry in future. 

Um... so by definition, can we say everything removed...?

File: app/Models/RemovalRequest.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    | FUNCTIONS
+    |--------------------------------------------------------------------------
+    */
+
+    /*
+    |--------------------------------------------------------------------------
+    | RELATIONS
+    |--------------------------------------------------------------------------
+    */
+
```
[the remark this answers is not in the document]
Comment (dan-tang-ssd) [reply app/Models/RemovalRequest.php:39]: As the corresponding organisation and user model will be deleted, there will be no model to refer to finally.
In this case, do we need to add relationship in model class...?

[the remark this answers is not in the document]
Comment (dave-mills) [reply app/Models/RemovalRequest.php:39]: I see your point, but I think the relationship would be long-term more maintainable. (It also means we don't keep data we shouldn't have (e.g. the org name), once the reomval request is completed and actioned.

[the remark this answers is not in the document]
Comment (dan-tang-ssd) [reply app/Models/RemovalRequest.php:39]: OK, I will revise program to add the relationships.

File: app/Http/Controllers/Admin/RemovalRequestCrudController.php [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        // TODO: remove project related records
+        // related tables: continent_project, country_project, project_region, portfolios, projects
+        if (count($projectIds) != 0) {
+            DB::statement('DELETE FROM continent_project WHERE project_id IN ' . $this->getSQLArray($projectIds));
+            DB::statement('DELETE FROM country_project WHERE project_id IN ' . $this->getSQLArray($projectIds));
+            DB::statement('DELETE FROM project_region WHERE project_id IN ' . $this->getSQLArray($projectIds));
+        }
+
+        DB::statement('DELETE FROM portfolios WHERE organisation_id = ' . $organisationId);
+        DB::statement('DELETE FROM projects WHERE organisation_id = ' . $organisationId);
```
[the remark this answers is not in the document]
Comment (dan-tang-ssd) [reply app/Http/Controllers/Admin/RemovalRequestCrudController.php]: Yes, I have just added ON DELETE constraints for most related tables.

One question, will it be possible to delete master records (users) when child records (organisation_members) are deleted. I cannot find out how to do this, so currently I still get all user Ids of an organisation then DELETE them via RAW SQL.

File: database/migrations/2023_05_16_100239_add_on_delete_for_organisation_related_tables.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        DB::statement("ALTER TABLE continent_project ADD CONSTRAINT fk_conp_projects FOREIGN KEY (project_id) REFERENCES projects(id) ON DELETE CASCADE;");
+        DB::statement("ALTER TABLE country_project ADD CONSTRAINT fk_coup_projects FOREIGN KEY (project_id) REFERENCES projects(id) ON DELETE CASCADE;");
+        DB::statement("ALTER TABLE project_region ADD CONSTRAINT fk_pr_projects FOREIGN KEY (project_id) REFERENCES projects(id) ON DELETE CASCADE;");
+
+        // add ON DELETE for organisation related tables
+        DB::statement("ALTER TABLE portfolios ADD CONSTRAINT fk_port_organisations FOREIGN KEY (organisation_id) REFERENCES organisations(id) ON DELETE CASCADE;");
+        DB::statement("ALTER TABLE projects ADD CONSTRAINT fk_pro_organisations FOREIGN KEY (organisation_id) REFERENCES organisations(id) ON DELETE CASCADE;");
+        DB::statement("ALTER TABLE organisation_members ADD CONSTRAINT fk_om_organisations FOREIGN KEY (organisation_id) REFERENCES organisations(id) ON DELETE CASCADE;");
+
+        // Question: Is there any way to delete master records (users) when child records (organisaiton_members) are deleted...?
```
Comment (dave-mills) [inline database/migrations/2023_05_16_100239_add_on_delete_for_organisation_related_tables.php:38]: Not that I know of! A good question... 

I think it's probably fine as-is, though. No need to add something complicated to pro-actively remove those users.

**Review by dave-mills — APPROVED** [review]
Looks great - and the extra constraints + on-delete triggers at the database level is a good solution to clearing out all the organisation's related entries. 👍

**Pull request merged**